Source/WebKit/UIProcess/glib/FenceMonitor.cpp

```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2024 Igalia S.L.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "FenceMonitor.h"
+
+#if PLATFORM(GTK) || (PLATFORM(WPE) && ENABLE(WPE_PLATFORM))
+#include <glib-unix.h>
+
+#if PLATFORM(GTK)
+#include <gtk/gtk.h>
+#endif
+
+namespace WebKit {
+
+FenceMonitor::FenceMonitor(Function<void()>&& callback)
+    : m_callback(WTFMove(callback))
+{
+}
+
+FenceMonitor::~FenceMonitor()
+{
+    if (m_source)
+        g_source_destroy(m_source.get());
+}
+
+struct FenceSource {
+    static GSourceFuncs sourceFuncs;
+
+    ~FenceSource()
+    {
+        if (tag)
+            g_source_remove_unix_fd(&base, tag);
+    }
+
+    GSource base;
+    UnixFileDescriptor fd;
+    gpointer tag { nullptr };
+};
+
+GSourceFuncs FenceSource::sourceFuncs = {
+    nullptr, // prepare
+    nullptr, // check
+    // dispatch
+    [](GSource* base, GSourceFunc callback, gpointer userData) -> gboolean
+    {
+        auto& source = *reinterpret_cast<FenceSource*>(base);
+        g_source_remove_unix_fd(&source.base, source.tag);
+        source.tag = nullptr;
+        source.fd = { };
+
+        callback(userData);
+        return G_SOURCE_CONTINUE;
+    },
```

Comment:
I think here it needs to check if the file descriptor is readable. Here's an example that does mostly the same: https://gitlab.gnome.org/GNOME/mutter/-/blob/main/src/wayland/meta-wayland-dma-buf.c?ref_type=heads#L868-926

Reply:
That was my inspiration :-) but here we only have one fd, so if the source is dispatched is because the fd became readable, no?

Reply:
I don't really know if the only event this FD can emit is when becoming readable...

Reply:
Reading it again, I think this part of the patch is perfectly fine as is already. I don't think it needs change.